src/lib/scheduler/operator_task.cpp

```diff
@@ -61,9 +57,25 @@ std::string OperatorTask::description() const {
 
 std::pair<std::vector<std::shared_ptr<AbstractTask>>, std::shared_ptr<OperatorTask>>
 OperatorTask::make_tasks_from_operator(const std::shared_ptr<AbstractOperator>& op) {
-  std::unordered_set<std::shared_ptr<OperatorTask>> operator_tasks_set;
+  auto operator_tasks_set = std::unordered_set<std::shared_ptr<OperatorTask>>{};
   const auto& root_operator_task = add_operator_tasks_recursively(op, operator_tasks_set);
 
+  // Ensure the task graph is acyclic, i.e., no task is any (n-th) successor of itself. Tasks in cycles would end up in
+  // a deadlock during execution, mutually waiting for the other tasks' execution. Even if the tasks are never executed,
+  // cycles create memory leaks since tasks hold shared pointers to their predecessors.
+  if constexpr (HYRISE_DEBUG) {
+    visit_tasks(root_operator_task, [](const auto& task) {
+      for (const auto& direct_successor : task->successors()) {
+        visit_tasks_upwards(direct_successor, [&](const auto& successor) {
+          Assert(task != successor, "Task graph contains a cycle.");
```

Comment:
Is equality check sufficient here (assuming it checks for pointer equality)? Can't we have cycles with copied nodes as well?

Reply:
We don't have such thing as task equality beyond mere identity. However, if we start copying tasks, we can also just stop developing a working system. Or do you think of `&task != &successor`?